sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py

```diff
@@ -34,23 +34,28 @@ def test_binomial_deviance():
     assert_almost_equal(bd(np.array([1.0, 0.0, 0.0]),
                            np.array([100.0, -100.0, -100.0])), 0)
 
-    # check if same results as alternative definition of deviance (from ESLII)
-    def alt_dev(y, pred):
-        return np.mean(np.logaddexp(0.0, -2.0 * (2.0 * y - 1) * pred))
+    # check if same results as alternative definition of deviance, from ESLII
+    # Eq. (10.18): -loglike = log(1 + exp(-2*z*f))
+    # Note:
+    # - We use y = {0, 1}, ESL (10.18) uses z in {-1, 1}, hence y=2*y-1
+    # - ESL 2*f = pred_raw, hence the factor 2 of ESL disappears.
+    # - Deviance = -2*loglike + .., hence a factor of 2 in front.
+    def alt_dev(y, raw_pred):
+        z = 2 * y - 1
+        return 2 * np.mean(np.log(1 + np.exp(-z * raw_pred)))
 
-    test_data = [(np.array([1.0, 1.0, 1.0]), np.array([100.0, 100.0, 100.0])),
-                 (np.array([0.0, 0.0, 0.0]), np.array([100.0, 100.0, 100.0])),
-                 (np.array([0.0, 0.0, 0.0]),
-                  np.array([-100.0, -100.0, -100.0])),
-                 (np.array([1.0, 1.0, 1.0]),
-                  np.array([-100.0, -100.0, -100.0]))]
+    test_data = product(
+        (np.array([0.0, 0.0, 0.0]), np.array([1.0, 1.0, 1.0])),
+        (np.array([-5.0, -5.0, -5.0]), np.array([3.0, 3.0, 3.0])))
 
     for datum in test_data:
         assert_almost_equal(bd(*datum), alt_dev(*datum))
 
-    # check the gradient against the
-    def alt_ng(y, pred):
-        return (2 * y - 1) / (1 + np.exp(2 * (2 * y - 1) * pred))
+    # check the negative gradient against altenative formula from ESLII
+    # Note: negative_gradient is half the negative gradient.
```

Comment:
I read this note twice and yet I still spend 5 minutes scratching my head wondering why this was missing a 2.
(and a minus sign.)
:facepalm:

Reply:
😆